episodes/12-fire-forget-tasks.md

```diff
@@ -0,0 +1,392 @@
+---
+title: "Fire-and-forget tasks"
+teaching: 60
+exercises: 30
+---
+
+:::::::::::::::::::::::::::::::::::::: questions
+- "How do execute work in parallel?"
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: objectives
+- "First objective."
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+A Chapel program always start as a single main thread. You can then start concurrent tasks with the `begin`
+statement. A task spawned by the `begin` statement will run in a different thread while the main thread
+continues its normal execution. Consider the following example:
+
+```chpl
+var x=0;
+
+writeln("This is the main thread starting first task");
+begin
+{
+  var c=0;
+  while c<100
+  {
+    c+=1;
+    writeln('thread 1: ',x+c);
+  }
+}
+
+writeln("This is the main thread starting second task");
+begin
+{
+  var c=0;
+  while c<100
+  {
+    c+=1;
+    writeln('thread 2: ',x+c);
+  }
+}
+
+writeln('this is main thread, I am done...');
+```
+
+```bash
+>> chpl begin_example.chpl -o begin_example
+>> ./begin_example
+```
+
+```output
+This is the main thread starting first task
+This is the main thread starting second task
+this is main thread, I am done...
+thread 2: 1
+thread 2: 2
+thread 2: 3
+thread 2: 4
+thread 2: 5
+thread 2: 6
+thread 2: 7
+thread 2: 8
+thread 2: 9
+thread 2: 10
+thread 2: 11
+thread 2: 12
+thread 2: 13
+thread 1: 1
+thread 2: 14
+thread 1: 2
+thread 2: 15
+...
+thread 2: 99
+thread 1: 97
+thread 2: 100
+thread 1: 98
+thread 1: 99
+thread 1: 100
+```
+
+As you can see the order of the output is not what we would expected, and actually it is completely
+unpredictable. This is a well known effect of concurrent tasks accessing the same shared resource at the same
+time (in this case the screen); the system decides in which order the tasks could write to the screen.
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Discussion
```

Comment:
This "discussion" imposes a high cognitive load on the instructor. Consider refactoring as a multiple-choice exercise.

Reply:
This is a good idea! I converted this discussion into two multiple-choice exercises and provided the solution for each. Resolved through a commit.